src/system/SystemPacketBuffer.cpp

```diff
@@ -541,24 +523,66 @@ PacketBufferHandle PacketBufferHandle::CloneData()
         // We do not clone an entire chain.
         return PacketBufferHandle();
     }
+    return NewWithData(mBuffer->Start(), mBuffer->DataLength());
+}
+
+#if CHIP_SYSTEM_CONFIG_USE_LWIP && LWIP_PBUF_FROM_CUSTOM_POOLS
+
+void PacketBufferHandle::RightSizeForLwIPCustomPools()
+{
+    PacketBuffer * lNewPacket = static_cast<PacketBuffer *>(pbuf_rightsize((struct pbuf *) mBuffer, -1));
+    if (lNewPacket != mBuffer)
+    {
+        mBuffer = lNewPacket;
+        SYSTEM_STATS_UPDATE_LWIP_PBUF_COUNTS();
+        ChipLogProgress(chipSystemLayer, "PacketBuffer: RightSize Copied");
+    }
+}
+
+#elif CHIP_SYSTEM_CONFIG_PACKETBUFFER_MAXALLOC == 0
 
-    PacketBufferHandle other = PacketBuffer::New();
-    if (other.IsNull())
+// Number of unused bytes below which \c RightSizeForMemoryAlloc() won't bother reallocating.
+constexpr uint16_t kRightSizingThreshold = 16;
+
+void PacketBufferHandle::RightSizeForMemoryAlloc()
+{
+    // Require a single buffer with no other references.
+    if ((mBuffer == nullptr) || (mBuffer->next != nullptr) || (mBuffer->ref != 1))
     {
-        return other;
+        return;
     }
 
-    if (other->AvailableDataLength() < mBuffer->DataLength())
+    // Reallocate only if enough space will be saved.
+    uint8_t * const start = reinterpret_cast<uint8_t *>(mBuffer) + CHIP_SYSTEM_PACKETBUFFER_HEADER_SIZE;
+    uint8_t * const payload = reinterpret_cast<uint8_t *>(mBuffer->payload);
+    const uint16_t usedSize = static_cast<uint16_t>(payload - start + mBuffer->len);
+    if (usedSize + kRightSizingThreshold > mBuffer->alloc_size)
     {
-        other = nullptr;
-        return other;
+        return;
     }
 
-    memcpy(other->Start(), mBuffer->Start(), mBuffer->DataLength());
-    other->SetDataLength(mBuffer->DataLength());
+    const uint16_t blockSize = usedSize + CHIP_SYSTEM_PACKETBUFFER_HEADER_SIZE;
+    PacketBuffer * newBuffer = reinterpret_cast<PacketBuffer *>(chip::Platform::MemoryAlloc(blockSize));
+    if (newBuffer == nullptr)
+    {
+        ChipLogError(chipSystemLayer, "PacketBuffer: pool EMPTY.");
+        return;
+    }
 
-    return other;
+    uint8_t * const newStart = reinterpret_cast<uint8_t *>(newBuffer) + CHIP_SYSTEM_PACKETBUFFER_HEADER_SIZE;
+    newBuffer->next = nullptr;
+    newBuffer->payload = newStart + (payload - start);
+    newBuffer->tot_len = mBuffer->tot_len;
+    newBuffer->len = mBuffer->len;
+    newBuffer->ref = 1;
+    newBuffer->alloc_size = static_cast<uint16_t>(usedSize);
+    memcpy(reinterpret_cast<uint8_t *>(newBuffer) + CHIP_SYSTEM_PACKETBUFFER_HEADER_SIZE, start, usedSize);
```

Comment:
This doesn't follow the chain. Rightsizing is to get a single linear buffer of correct chain from a single PacketBuffer of size > optiomal OR from a chain. This would need to take in account the chain, otherwise this may copy past the end of the buffer. I am surprised to see `->len` being used directly, rather than looking at the entire chain (e.g. equivalent of TotalLength)

Reply:
Yes. I thought I had written that, but evidently didn't (or misplaced it on another branch). Will move to draft until I add it.

Reply:
Memory has been refreshed. The LwIP implementation also does not support chains, so I added the comment to call `CompactHead()` first for that case. If `RightSize()` itself needs to handle chains, I'd rather land that separately.